### ISSUE TITLE
Runtime Error encountered when using the Geneos utility to start a gateway

### ISSUE THREAD

**Issue by gadvincula** [ITRS-Group/cordial#148]
Background:

1. I installed a gateway in a client's UAT server using the Geneos utility.
2. a runtime error was encountered when I try to start the gateway I just created
3. Operating system of the UAT server is RHEL 7.9

Full Error Details:

$ geneos start gateway UAT_GATEWAY_TEST
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x89e357]

goroutine 1 [running]:
github.com/itrs-group/cordial/pkg/host.(*Local).Username(0xb468e0?)
        /app/cordial/pkg/host/local.go:51 +0x17
github.com/itrs-group/cordial/tools/geneos/internal/instance.Start({0xb468e0?, 0xc00007a0a0})
        /app/cordial/tools/geneos/internal/instance/start.go:42 +0xce
github.com/itrs-group/cordial/tools/geneos/cmd.Start.func1({0xb468e0?, 0xc00007a0a0?}, {0x1b?, 0xc0002078d0?, 0x97f660?})
        /app/cordial/tools/geneos/cmd/start.go:64 +0x25
github.com/itrs-group/cordial/tools/geneos/internal/instance.ForAll(0xc000012a08?, {0xa458c4, 0x3}, 0xaaaab0, {0xc000025300, 0x1, 0x4}, {0x                                                                                                  0, 0x0, 0x0})
        /app/cordial/tools/geneos/internal/instance/instance.go:465 +0x2f7
github.com/itrs-group/cordial/tools/geneos/cmd.Start(0xe70780?, 0x0, {0xc000025300, 0x1, 0x4}, {0x0, 0x0, 0x0})
        /app/cordial/tools/geneos/cmd/start.go:63 +0x90
github.com/itrs-group/cordial/tools/geneos/cmd.glob..func28(0xe70780?, {0xc00003a220?, 0x2?, 0x2?})
        /app/cordial/tools/geneos/cmd/start.go:58 +0x39
github.com/spf13/cobra.(*Command).execute(0xe70780, {0xc00003a1c0, 0x2, 0x2})
        [/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:940](mailto:/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:940) +0x862
github.com/spf13/cobra.(*Command).ExecuteC(0xe6f920)
        [/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068](mailto:/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068) +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
        [/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992](mailto:/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992)
github.com/itrs-group/cordial/tools/geneos/cmd.Execute()
        /app/cordial/tools/geneos/cmd/root.go:231 +0x25
main.main()
        /app/cordial/tools/geneos/main.go:72 +0xfd

As part of troubleshooting, I tried to start the gateway manually and I was able to start it.

$ ./gateway2.linux_64 -setup /home/uat/ugeneos2/bin/geneos/gateway/gateways/UAT_GATEWAY_TEST/gateway.setup.xml -licd-host geneos-licd2.jpmchase.net -licd-port 6600
2023-06-23 02:52:58.371-0500 ***********************************************************
2023-06-23 02:52:58.371-0500 Geneos EMF2 Gateway GA6.3.1-230427 (c) 1995-2023 ITRS Group
2023-06-23 02:52:58.371-0500 ***********************************************************
2023-06-23 02:52:58.373-0500  INFO: Build type: linux_64:Release
2023-06-23 02:52:58.373-0500  INFO: Running from: /home/uat/ugeneos2/bin/geneos/packages/gateway/6.3.1
2023-06-23 02:52:58.373-0500  INFO: User/uid/gid: ugeneos2/92170/11536
2023-06-23 02:52:58.373-0500  INFO: Current path: /usr/lib64/qt-3.3/bin:/usr/local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/opt/puppetlabs/bin:/home/uat/ugeneos2/bin/geneos
2023-06-23 02:52:58.373-0500  INFO: LD_LIBRARY_PATH has not been set
2023-06-23 02:52:58.373-0500  INFO: Gateway Gateway statistics collection is off.
2023-06-23 02:52:58.376-0500  INFO: SSLContext TLS version 1.0 and upwards accepted for secure connections.
2023-06-23 02:52:58.383-0500  INFO: CURL libcurl version: 7.69.1-DEV
2023-06-23 02:52:58.383-0500  INFO: CURL ssl version: OpenSSL/1.1.1q
2023-06-23 02:52:58.383-0500  INFO: CURL Support for: SSL NTLM IPv6 SPNEGO;  Protocols: dict file ftp ftps gopher http https imap imaps pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
2023-06-23 02:52:58.444-0500  INFO: LicenceManager Connecting to Licence Daemon ("***removing this info - Gian")
2023-06-23 02:52:58.444-0500  INFO: HostInfo: "***removing this info - Gian", Version #1 SMP Mon Dec 19 10:44:06 UTC 2022, Platform Linux, Release 3.10.0-1160.83.1.el7.x86_64, MachinS Linux 3.10.0-1160.83.1.el7.x86_64
2023-06-23 02:52:58.555-0500  INFO: SetupManager MD5Hash:6a19a77b8a1cac5e197e5aaf0f00fe2e | Read from Cache : File:/home/uat/ugeneos2/bin/geneos/gateway/gateways/UAT_GATEWAY_TEST/gateway.setup.xml=>5984ecf4f30e6d823c428a863
2023-06-23 02:52:58.556-0500  INFO: SetupManager MD5Hash:be649c8620983b8e5eecffe2a3b798df | Read from Cache : File:/home/uat/ugeneos2/bin/geneos/gateway/gateways/UAT_GATEWAY_TEST/instance.setup.xml=>726e100d3d4e6482786667bee
2023-06-23 02:52:58.556-0500  INFO: SetupManager MD5Hash:6a19a77b8a1cac5e197e5aaf0f00fe2e | Read from File : File:/home/uat/ugeneos2/bin/geneos/gateway/gateways/UAT_GATEWAY_TEST/gateway.setup.xml
2023-06-23 02:52:58.556-0500  INFO: SetupManager MD5Hash:be649c8620983b8e5eecffe2a3b798df | Read from File : File:/home/uat/ugeneos2/bin/geneos/gateway/gateways/UAT_GATEWAY_TEST/instance.setup.xml
2023-06-23 02:52:58.561-0500  INFO: SetupManager Applying setup.
2023-06-23 02:52:58.561-0500  INFO: Gateway Listening on INSECURE port 7039
2023-06-23 02:52:58.561-0500  INFO: Gateway Unique id is 18694698
2023-06-23 02:52:58.561-0500  INFO: Gateway Unique id was generated using the gateway name.
2023-06-23 02:52:58.561-0500  INFO: Gateway Gateway name is UAT_GATEWAY_TEST
2023-06-23 02:52:58.561-0500  INFO: Gateway history rollover file limit set to 10
2023-06-23 02:52:58.561-0500  INFO: Log archive script is not set.
2023-06-23 02:52:58.561-0500  INFO: CONMANAGER sleepInTryHeartbeat is true
2023-06-23 02:52:58.561-0500  INFO: CONMANAGER tryHeartbeatIterations is set to 10
2023-06-23 02:52:58.561-0500  INFO: CONMANAGER heartbeatInterval is set to 75
2023-06-23 02:52:58.561-0500  INFO: CONMANAGER connectWait time is set to 30
2023-06-23 02:52:58.566-0500  INFO: SchemaManager The working version is now GA6.3.1-230427
2023-06-23 02:52:58.567-0500  INFO: SelfAnnouncingProbeManager Self-announcing probes feature now enabled. Probe limit is unlimited.
2023-06-23 02:52:58.568-0500  INFO: RuleManager Using 0 rule worker threads
2023-06-23 02:52:58.569-0500  INFO: SetupManager MD5Hash:6a19a77b8a1cac5e197e5aaf0f00fe2e | Read from File : File:/home/uat/ugeneos2/bin/geneos/gateway/gateways/UAT_GATEWAY_TEST/gateway.setup.xml
2023-06-23 02:52:58.569-0500  INFO: SetupManager MD5Hash:6a19a77b8a1cac5e197e5aaf0f00fe2e | Write to Cache : File:/home/uat/ugeneos2/bin/geneos/gateway/gateways/UAT_GATEWAY_TEST/gateway.setup.xml=>5984ecf4f30e6d823c428a865
2023-06-23 02:52:58.569-0500  INFO: SetupManager MD5Hash:be649c8620983b8e5eecffe2a3b798df | Read from File : File:/home/uat/ugeneos2/bin/geneos/gateway/gateways/UAT_GATEWAY_TEST/instance.setup.xml
2023-06-23 02:52:58.569-0500  INFO: SetupManager MD5Hash:be649c8620983b8e5eecffe2a3b798df | Write to Cache : File:/home/uat/ugeneos2/bin/geneos/gateway/gateways/UAT_GATEWAY_TEST/instance.setup.xml=>726e100d3d4e6482786667bde
2023-06-23 02:52:58.570-0500  INFO: SetupManager Finished applying setup.
2023-06-23 02:52:58.620-0500  WARN: GatewayLicenceManager Gateway is not licensed.
2023-06-23 02:52:58.670-0500  INFO: HOSTRESOLVER Timeout set to  15 seconds
2023-06-23 02:52:58.670-0500  INFO: HOSTRESOLVER resolver using  gethostbyname_r()
2023-06-23 02:52:58.670-0500  INFO: HOSTRESOLVER cache status is Enabled
2023-06-23 02:52:58.671-0500  INFO: HOSTRESOLVER operating mode is THREADED
2023-06-23 02:52:58.671-0500  INFO: HOSTRESOLVER trying to resolve "***removing this info - Gian"
2023-06-23 02:52:58.708-0500  INFO: HOSTRESOLVER Took 0 Seconds to resolve host ["***removing this info - Gian"]
2023-06-23 02:52:58.708-0500  INFO: HOSTRESOLVER host "***removing this info - Gian" resolved to "***removing this info - Gian"
2023-06-23 02:52:58.797-0500  INFO: HOSTRESOLVER Address info service: Max acceptable DNS lookup time 15 seconds
2023-06-23 02:52:58.797-0500  INFO: HOSTRESOLVER Duration DNS lookup suspension on failure, 300 seconds
2023-06-23 02:52:59.523-0500  INFO: LicenceManager Connected to Licensing Daemon
2023-06-23 02:52:59.823-0500  INFO: GatewaySystem Gateway is licenced.
2023-06-23 02:52:59.823-0500  INFO: RoleManager Gateway is now running as Stand Alone
2023-06-23 02:52:59.823-0500  INFO: RoleManager Gateway switching to Active state
2023-06-23 02:52:59.826-0500  INFO: RoleManager Gateway is now Active
2023-06-23 02:53:00.131-0500  INFO: SelfAnnouncingProbeManager Self Announcing Probes are licensed.
^C2023-06-23 02:54:16.151-0500  INFO: Gateway Signal INT received - terminating
2023-06-23 02:54:16.225-0500  INFO: Gateway Exiting.
2023-06-23 02:54:16.227-0500  INFO: DatabaseLogging Disabled

Has anyone experienced this error before? As an FYI, I performed the same work of installing the Geneos utility, gateway installation and started the gateway on the client's DEV server and it's working fine. That server is running the same RHEL 7.9 OS.

**Comment by gadvincula** [comment]
I am comparing the difference between the DEV and UAT server of the client.

So far the major difference I noticed was the geneos account I used in DEV can be found in the server's /etc/passwd file while in the UAT server it's not present.

**Comment by pgalbavy-itrs** [comment]
Thanks for the report. The panic is because the code assumed,  wrongly, that querying the current user will always succeed.

I am updating the code in this case and other similar assumed usages.

The normal build of `geneos` is statically linked and so cannot take part in PAM authentication lookups where the user may be in an external directory. There was a build using Centos 7 in the release, it is just not in the main downloads - that should resolve the issue on typical Debian systems - download the .tar.gz and look in the bin directory.

Meanwhile I will apply the fixes (which fall back to environment variables for USER and HOME) in the next patch release shortly.

**Comment by pgalbavy-itrs** [comment]
FYI This is now fixed in v1.6.5 - https://github.com/ITRS-Group/cordial/releases/tag/v1.6.5 - and I have pulled out the "centos7" binary which will also do PAM lookups, if you want to use that

**Comment by gadvincula** [comment]
Thank you Peter. I will use v1.6.5 and test it today.

**Comment by gadvincula** [comment]
<img width="677" alt="ss" src="https://github.com/ITRS-Group/cordial/assets/114900919/311fae90-2763-431b-b61d-89a933566a3c">

I used the one with the filename "geneos" only and so far its working on my local test virtual machine.

How do you use the one with the filename "geneos.centos7-x86_64"? It doesn't seem to be a binary that's ready to be used similar to "geneos"

**Comment by gadvincula** [comment]
Another observation. After using this new version of the Geneos utility, there seems to be an issue when you change the port of your gateway using the "geneos set gateway <gateway_name> port=xxxx" command.

My test gateway initially got assigned the default 7039 port and I was able to start my gateway with it. 

-bash-4.2$ geneos ls
Type     Names         Host  Flag       Port  Version            Home
gateway  test_gateway  A     localhost  7039  active_prod:6.2.0  /home/geneos/gateway/gateways/test_gateway

However, I tried updating my port from 7039 to 7040 using the "geneos set gateway <gateway_name> port=xxxx" command." 

-bash-4.2$ geneos set gateway test_gateway port=7040

The command seem to update the port when you run the "geneos ps" or "geneos show gateway" command.

-bash-4.2$ geneos ls
Type     Names         Host  Flag       Port  Version            Home
gateway  test_gateway  A     localhost  7040  active_prod:6.2.0  /home/geneos/gateway/gateways/test_gateway

-bash-4.2$ geneos show gateway test_gateway | grep port
            "port": "7040",

However, the <listenPort> in the instance.setup.xml doesn't get updated.

-bash-4.2$ pwd
/home/geneos/gateway/gateways/test_gateway
-bash-4.2$ grep -i port instance.setup.xml
                <listenPorts>
                                <listenPort>7039</listenPort>
                </listenPorts>

I was able to find this out as I was not able to start my gateway after updating my port from 7039 to 7040.
